### PR TITLE
[READY] - Reworking default.nix and adding overlay

### PR DESCRIPTION
**Pull request by sarcasticadmin**
## Description
We should be leveraging an overlay and organizing our nix packages in a more clear manner. This should start us down that path

## Acceptance Criteria
* Images still build as expected
* Anyone can consume overlay.nix from this repo

## Testing

Using the new overlay. I had to leverage my branch but once we merge this into `master` we can leverage it via what I added in the README:

A sample `shell.nix`:

```nix
let
  nix-garage = builtins.fetchGit {
    url = "https://github.com/sarcasticadmin/nix-garage";
    ref = "sa/gl-issue-309";
  };
  garage-overlay = import (nix-garage.outPath + "/overlay.nix");
  nixpkgs = import <nixpkgs> { overlays = [ garage-overlay ]; };
in
  with nixpkgs;
  mkShell {
    buildInputs = [
        flasksample
    ];
  }
```

Then drop into the nix-shell and run the `flasksample` pkg:

```bash
$ nix-shell
[nix-shell] $ export FLASK_APP=flasksample.app
[nix-shell] $ flask run
* Serving Flask app "flasksample.app"
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

Then making sure that we can still build our images and reproducibility is still intacted:

```
$ for currImgPath in $(ls -d imgs/*/); do  ./publish-imgs $(basename $currImgPath); done
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
{                                                                                              
    "Name": "docker.io/nebulaworks/awsutils",                                                                                                                
    "Digest": "sha256:d22b87ad666430779b782c307b5afc52968995b860d7a77656a93a3f413ca14a",
    "RepoTags": [                                                                       
        "0afjz2x6jzlqbyp6hlsn7xx1j4v595xj",                                                                       
        "dz783iaashdby6phqf9pcydxsw1bbjq0",                    
        "ib0v6zkv2qqaidwda7sp64pa0jfc0vbs"                                            
    ],                                
    "Created": "1970-01-01T00:00:01Z",                                           
    "DockerVersion": "",
    "Labels": {                                                                               
        "com.nebulaworks.packages": "awscli:1.17.13,bash:4.4,coreutils:8.31,curl:7.68.0,jq:1.6,nss-cacert:3.49.2",
        "org.opencontainers.image.authors": "Nebulaworks Inc.",                       
        "org.opencontainers.image.source": "https://github.com/Nebulaworks/nix-garage"
    },                      
    "Architecture": "amd64",                                                                   
    "Os": "linux",
    "Layers": [
         "sha256:2c8d5e5a38c0d8ae54c419a59a7df5d6106f66903df4dd26059e5a5d89a30f33"                                                                                                   
    ],
    "Env": [
        "PATH=/bin/",
        "SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt"
    ]
}
copying path '/nix/store/84qi3iba9m2xs2jm40kccvh0hyakkf2s-source' from 'https://cache.nixos.org'...
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
{
    "Name": "docker.io/nebulaworks/helmsman-aws",
    "Digest": "sha256:2cab619b7e43b30eb8384b9104b19499e4b0ac39c4347ae77ec5b3c65007c103",
    "RepoTags": [
        "0r4rskk3b5han0zzncj67sjhkm7lip1w",
        "2vil2iz4gz0lmcqsg92kxylhh4mds6jv",
        "72rs42ar65syxrgq6w0qyv98n97zkr6n",
        "7ar2sigpxxyll06yvgan428qn74k0vqm",
        "7hl8p275d7f3gcnam52s7hg1d307a2ik",
        "axk5j5bhv0sv9ys1xj4bmhc34kfw5jkn",
        "ibx9pd0yzvspf6gczmdmmhmam9sr25wr",
        "pbj9q2qyw511wfy18g8lz9jw2mpc5qvw",
        "sb47bkwd854ws9dc4gzsvnln06pidxjf",
        "xym39zji91ga5rbw6p8xbwy5dcbizbn3"
    ],
    "Created": "1970-01-01T00:00:01Z",
    "DockerVersion": "",
    "Labels": {
        "com.nebulaworks.packages": "awscli:1.17.13,bash:4.4,coreutils:8.31,helm-diff:3.1.1,helm:3.2.1,helmsman:3.3.0,jq:1.6,kubectl:1.18.1,nss-cacert:3.52",
        "org.opencontainers.image.authors": "Nebulaworks Inc.",
        "org.opencontainers.image.source": "https://github.com/Nebulaworks/nix-garage"
    },
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:b59069cadaedf5c0ebef846dcc248bb64a9ee16b972fe6c8d6769efec56a1795"
    ],
    "Env": [
        "PATH=/bin/",
        "HELM_PLUGINS=/share/helm/plugins/"
    ]
}
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
{
    "Name": "docker.io/nebulaworks/magic-wormhole-mailbox",
    "Digest": "sha256:6bb9e8430553918346aa3e602727d5ef657171c347749b01f52bf77e57492c56",
    "RepoTags": [
        "msahb86b470r5636v8imzs8sghgzg38b"
    ],
    "Created": "1970-01-01T00:00:01Z",
    "DockerVersion": "",
    "Labels": {
        "com.nebulaworks.packages": "bash:4.4,coreutils:8.31,procps:3.3.16,python3:3.7.6-env",
        "org.opencontainers.image.authors": "Nebulaworks Inc.",
        "org.opencontainers.image.source": "https://github.com/Nebulaworks/nix-garage"
    },
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:2dd20c82cb559ffbb4a4b185619d524dee48b3f3b34af40349bf98aca07874cc"
    ],
    "Env": [
        "PATH=/bin/"
    ]
}
warning: you did not specify '--add-root'; the result might be removed by the garbage collector
{
    "Name": "docker.io/nebulaworks/pki-validator",
    "Digest": "sha256:9be4e1655ccb7b906f9c0109394ad6bb733634c3e4862c6866f8cf19d5829c71",
    "RepoTags": [
        "12x2y4s04wqd7vzblyknmb4i9wlrbyg7"
    ],
    "Created": "1970-01-01T00:00:01Z",
    "DockerVersion": "",
    "Labels": {
        "com.nebulaworks.packages": "bash:4.4,coreutils:8.31,gawk:5.0.1,gnugrep:3.4,gnumake:4.2.1,openssl:1.1.1d",
        "org.opencontainers.image.authors": "Nebulaworks Inc.",
        "org.opencontainers.image.source": "https://github.com/Nebulaworks/nix-garage"
    },
    "Architecture": "amd64",
    "Os": "linux",
    "Layers": [
        "sha256:e6ac3abb53a8af091b4c040681b8615355e84bdf1ac92a57c87312253ecce851"
    ],
    "Env": [
        "PATH=/bin/",
        "SSL_CERT_FILE=/etc/ssl/certs/ca-bundle.crt"
    ]
}
```
No new images 😄 ! And we cant completely validate CI since its reading its configuration from `.github` dir off `master` which makes some assumes that Ive corrected for these changes.

### CODE REVIEW

**Comment by sarcasticadmin** [comment]
publish

**Comment by sarcasticadmin** [comment]
publish

File: .github/workflows/publish_imgs.yml
Comment (sarcasticadmin) [inline .github/workflows/publish_imgs.yml:120]: We dont need to build the images each time, lets leave that up to `publish-imgs` to decide

File: .github/workflows/publish_imgs.yml
Comment (sarcasticadmin) [inline .github/workflows/publish_imgs.yml:87]: Removing this since I think it was only here for troubleshooting

File: publish-imgs
Comment (sarcasticadmin) [inline publish-imgs:20]: Lets build the image only when its not in the registry

**Comment by sarcasticadmin** [comment]
@Nebulaworks/approvers ready to go here!

File: .github/workflows/publish_imgs.yml
Comment (jamesnebulaworks) [reply .github/workflows/publish_imgs.yml:120]: awesome!

File: publish-imgs
Comment (jamesnebulaworks) [reply publish-imgs:20]: thanks for adding this in here so we dont have to build unnecessarily 👍

**Comment by jamesnebulaworks** [comment]
@sarcasticadmin everything looks good to me rob! super slick concept with the overlay. i'll wait for @maishiroma to chime in before merging 🙂

File: .github/workflows/publish_imgs.yml
Comment (maishiroma) [reply .github/workflows/publish_imgs.yml:87]: Yeah this was mostly to help with debugging. If anything, we could comment it out, but I think it's fine to remove it from the yml itself.

File: .github/workflows/publish_imgs.yml
Comment (maishiroma) [reply .github/workflows/publish_imgs.yml:120]: So my understanding is that the script, `publish-imgs` will take care of that? I like that approach much better. 😃 .

File: README.md
Comment (maishiroma) [inline README.md:29]: So if I am understanding this correctly, this is like a docker image layers where you can layer packages onto one another to combine them all into one package. The big difference is that you can control if you want a package defined in a lower layer to override what was already put in the overlay? It also allows for packages to _reuse_ dependencies that can be shared between packages?

File: publish-imgs
Comment (maishiroma) [reply publish-imgs:20]: Could @sarcasticadmin clarify on the syntax here `imgs.$1`? I know we are passing in the image name through the CI, but why the `imgs.` before it? Is that a nix specific syntax nuance?

**Review by maishiroma — COMMENTED** [review]
This looks good, I just have a few questions that I would like some quick clarification, primarily on the `imgs.$1` in the `publish-imgs`. Other than that, I like what we did here especially with looking into overlays, never knew about those in `nix`. 😃

File: .github/workflows/publish_imgs.yml
Comment (sarcasticadmin) [reply .github/workflows/publish_imgs.yml:120]: Exactly!

File: README.md
Comment (sarcasticadmin) [reply README.md:29]: It allows us to incorporate our own pkgs with the nixpkg tree or override specific existing pkgs in nixpkgs with custom configuration. Check out the wiki on this: https://nixos.wiki/wiki/Overlays for some other examples.

But the main idea is reusability. You can add this nix-garage overlay to your local nix environment and install things just like you would anything from the upstream nixpkg repo. Where this gets really cool is when we need to mix in N overlays from various sources 🙂  but well get to that later

File: publish-imgs
Comment (sarcasticadmin) [reply publish-imgs:20]: The syntax is due to the `import` at the top-level here: https://github.com/Nebulaworks/nix-garage/pull/19/files#diff-5712e736e0de6ba170577f8472c398e9R7 all things defined in `imgs/default.nix` will now have an attribute prefix of `imgs.<pkg>`. This is separate from the `overlay.nix` thats defined earlier so sorry if theres confusion there. I didnt feel that it would be useful for use to have the docker images included in the overlay.

**Review by sarcasticadmin — COMMENTED** [review]
@jamesnebulaworks @maishiroma really appreciate the feedback from both of you! Let me know if my comments addressed your questions.

File: publish-imgs
Comment (maishiroma) [reply publish-imgs:20]: Ah I see the relationship now that makes a lot more sense, that's a nice way to easily tie in the entire image directory into the `default.nix` derivation.

File: README.md
Comment (maishiroma) [reply README.md:29]: Ahhhh I see it's basically "inheriting" the base nixpkgs set, but adding in our own. That's really cool reminds me a lot of inheritance in OOP (I like those languages ahaha 😆 ). Except this sounds a lot more flexible than inheritance, so I'm all in for that. I'll have to read up more on overlays!

**Review by maishiroma — APPROVED** [review]
Thanks @sarcasticadmin for the clarification on my comments, that's really cool to see how the next iteration is looking really clean now. Thanks for getting this through! LGTM 🎉

**Pull request merged**
